### PR TITLE
WinUpdater: Quote path for args when relaunching Dolphin.

### PR DESCRIPTION
**Pull request by AdmiralCurtiss**
This should fix https://forums.dolphin-emu.org/Thread-recent-update-triggering-this-error-on-auto-update

I *think* just adding quotes on both ends is enough, because a valid path on Windows cannot contain quotes by itself.

### CODE REVIEW

**Comment by AdmiralCurtiss** [comment]
Fair enough.

**Comment by AdmiralCurtiss** [comment]
I did test this yesterday and it works as written, but yeah maybe some dumb Windows fallback is saving me here...

**Comment by AdmiralCurtiss** [comment]
Btw the way I'm reading [the MSDN doc](https://learn.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-createprocessw) is that the application name only needs to be quoted if either `lpApplicationName` or `lpCommandLine` is NULL (since that causes the non-NULL of them to be interpreted as the command line); and if neither is NULL then the `lpApplicationName` is the path to the application (which doesn't need quotes, based on the security remark at the bottom) and the `lpCommandLine` does need quotes since it doesn't know where to split the command line arguments otherwise.

**Comment by AdmiralCurtiss** [comment]
Okay, I tested this out under the following scenario taking into account the Security stuff on the MSDN page:

Clipboard0233

---

`c:\Test\a dir\a.exe` is the 'correct' executable that we want to launch. It's a simple C# WinForms application that prints 'Correct Executable', and in the textbox its own path, then its working directory, then its arguments.

Clipboard01

<details>
<summary>C# Code</summary><br>

```
namespace WinFormsApp1 {
	internal static class Program {
		/// <summary>
		///  The main entry point for the application.
		/// </summary>
		[STAThread]
		static void Main(string[] args) {
			// To customize application configuration such as set high DPI settings or default font,
			// see https://aka.ms/applicationconfiguration.
			ApplicationConfiguration.Initialize();
			Application.Run(new Form1(args));
		}
	}

	public partial class Form1 : Form {
		public Form1(string[] args = null) {
			InitializeComponent();

			textBox1.Text += System.Reflection.Assembly.GetExecutingAssembly().Location;
			textBox1.Text += "\r\n";
			textBox1.Text += Directory.GetCurrentDirectory();
			textBox1.Text += "\r\n";
			textBox1.Text += "\r\n";
			foreach (string s in args) {
				textBox1.Text += s + "\r\n";
			}
		}
	}
}
```

</details>

`c:\Test\a.exe` is largely identical but prints 'Wrong Executable' instead.

Clipboard04

---

We now attempt to launch this in a few different ways using `CreateProcessW()`:

<details>
<summary>Both parameters, both unquoted</summary><br>

```
#include <string>

#include <Windows.h>

int main(int argc, char** argv) {
	std::wstring unquoted = L"c:\\Test\\a dir\\a.exe";
	std::wstring quoted = L"\"" + unquoted + L"\"";

	std::wstring lpApplicationName = unquoted;
	std::wstring lpCommandLine = unquoted;

	STARTUPINFOW startup_info{ .cb = sizeof(startup_info) };
	PROCESS_INFORMATION process_info;
	BOOL rv = CreateProcessW(lpApplicationName.data(), lpCommandLine.data(), nullptr, nullptr, TRUE, 0, nullptr, nullptr, &startup_info, &process_info);
	printf("%s\n", rv ? "launch success" : "launch failed");
}
```

==> Launches correct executable (prints `launch success`) with:

```
c:\Test\a dir\a.dll
c:\Test

dir\a.exe
```

So we launched the correct file, but with the argument `dir\a.exe`. If we delete `c:\Test\a dir\a.exe` nothing gets launched (prints `launch failed`).

</details>


<details>
<summary>Both parameters, both quoted</summary><br>

```
#include <string>

#include <Windows.h>

int main(int argc, char** argv) {
	std::wstring unquoted = L"c:\\Test\\a dir\\a.exe";
	std::wstring quoted = L"\"" + unquoted + L"\"";

	std::wstring lpApplicationName = quoted;
	std::wstring lpCommandLine = quoted;

	STARTUPINFOW startup_info{ .cb = sizeof(startup_info) };
	PROCESS_INFORMATION process_info;
	BOOL rv = CreateProcessW(lpApplicationName.data(), lpCommandLine.data(), nullptr, nullptr, TRUE, 0, nullptr, nullptr, &startup_info, &process_info);
	printf("%s\n", rv ? "launch success" : "launch failed");
}
```

==> Does not launch anything. Prints `launch failed`.

</details>


<details>
<summary>Both parameters, lpApplicationName unquoted, lpCommandLine quoted</summary><br>

```
#include <string>

#include <Windows.h>

int main(int argc, char** argv) {
	std::wstring unquoted = L"c:\\Test\\a dir\\a.exe";
	std::wstring quoted = L"\"" + unquoted + L"\"";

	std::wstring lpApplicationName = unquoted;
	std::wstring lpCommandLine = quoted;

	STARTUPINFOW startup_info{ .cb = sizeof(startup_info) };
	PROCESS_INFORMATION process_info;
	BOOL rv = CreateProcessW(lpApplicationName.data(), lpCommandLine.data(), nullptr, nullptr, TRUE, 0, nullptr, nullptr, &startup_info, &process_info);
	printf("%s\n", rv ? "launch success" : "launch failed");
}
```

==> Launches correct executable (prints `launch success`) with:

```
c:\Test\a dir\a.dll
c:\Test
```

So the process launched and did not get any arguments, exactly what we want.

</details>


<details>
<summary>Both parameters, lpApplicationName quoted, lpCommandLine unquoted</summary><br>

```
#include <string>

#include <Windows.h>

int main(int argc, char** argv) {
	std::wstring unquoted = L"c:\\Test\\a dir\\a.exe";
	std::wstring quoted = L"\"" + unquoted + L"\"";

	std::wstring lpApplicationName = quoted;
	std::wstring lpCommandLine = unquoted;

	STARTUPINFOW startup_info{ .cb = sizeof(startup_info) };
	PROCESS_INFORMATION process_info;
	BOOL rv = CreateProcessW(lpApplicationName.data(), lpCommandLine.data(), nullptr, nullptr, TRUE, 0, nullptr, nullptr, &startup_info, &process_info);
	printf("%s\n", rv ? "launch success" : "launch failed");
}
```

==> Does not launch anything. Prints `launch failed`.

</details>

For completeness sake:

<details>
<summary>lpCommandLine only, unquoted</summary><br>

```
#include <string>

#include <Windows.h>

int main(int argc, char** argv) {
	std::wstring unquoted = L"c:\\Test\\a dir\\a.exe";
	std::wstring quoted = L"\"" + unquoted + L"\"";

	//std::wstring lpApplicationName;
	std::wstring lpCommandLine = unquoted;

	STARTUPINFOW startup_info{ .cb = sizeof(startup_info) };
	PROCESS_INFORMATION process_info;
	BOOL rv = CreateProcessW(NULL, lpCommandLine.data(), nullptr, nullptr, TRUE, 0, nullptr, nullptr, &startup_info, &process_info);
	printf("%s\n", rv ? "launch success" : "launch failed");
}
```

==> Launches the *wrong* executable (`launch success`) with:

```
c:\Test\a.dll
c:\Test

dir\a.exe
```

That's not too surprising, that's exactly what the MSDN article warns against doing.
</details>

<details>
<summary>lpApplicationName only, unquoted</summary><br>

```
#include <string>

#include <Windows.h>

int main(int argc, char** argv) {
	std::wstring unquoted = L"c:\\Test\\a dir\\a.exe";
	std::wstring quoted = L"\"" + unquoted + L"\"";

	std::wstring lpApplicationName = unquoted;
	//std::wstring lpCommandLine = unquoted;

	STARTUPINFOW startup_info{ .cb = sizeof(startup_info) };
	PROCESS_INFORMATION process_info;
	BOOL rv = CreateProcessW(lpApplicationName.data(), NULL, nullptr, nullptr, TRUE, 0, nullptr, nullptr, &startup_info, &process_info);
	printf("%s\n", rv ? "launch success" : "launch failed");
}
```

==> Launches the *correct* executable (`launch success`) with:

```
c:\Test\a dir\a.dll
c:\Test
```

This one actually surprises me! I would have expected this to behave the same as the lpCommandLine only variant based on the documentation.

If we delete the correct executable this fails. (`launch failed`)
</details>


<details>
<summary>lpApplicationName only, quoted</summary><br>

```
#include <string>

#include <Windows.h>

int main(int argc, char** argv) {
	std::wstring unquoted = L"c:\\Test\\a dir\\a.exe";
	std::wstring quoted = L"\"" + unquoted + L"\"";

	std::wstring lpApplicationName = quoted;
	//std::wstring lpCommandLine = unquoted;

	STARTUPINFOW startup_info{ .cb = sizeof(startup_info) };
	PROCESS_INFORMATION process_info;
	BOOL rv = CreateProcessW(lpApplicationName.data(), NULL, nullptr, nullptr, TRUE, 0, nullptr, nullptr, &startup_info, &process_info);
	printf("%s\n", rv ? "launch success" : "launch failed");
}
```

==> Fails to launch anything. (`launch failed`)

</details>

<details>
<summary>lpCommandLine only, quoted</summary><br>

```
#include <string>

#include <Windows.h>

int main(int argc, char** argv) {
	std::wstring unquoted = L"c:\\Test\\a dir\\a.exe";
	std::wstring quoted = L"\"" + unquoted + L"\"";

	//std::wstring lpApplicationName = quoted;
	std::wstring lpCommandLine = quoted;

	STARTUPINFOW startup_info{ .cb = sizeof(startup_info) };
	PROCESS_INFORMATION process_info;
	BOOL rv = CreateProcessW(NULL, lpCommandLine.data(), nullptr, nullptr, TRUE, 0, nullptr, nullptr, &startup_info, &process_info);
	printf("%s\n", rv ? "launch success" : "launch failed");
}
```

==> Launches the correct executable (`launch success`) with:

```
c:\Test\a dir\a.dll
c:\Test
```

If we delete the executable it fails. (`launch failed`)

</details>

**Comment by AdmiralCurtiss** [comment]
Based on this:

- Adding quotes to the first parameter is definitely wrong, it never launches anything if you do that.
- Adding quotes to the second parameter is correct.
- Leaving the second parameter NULL seems to behave exactly like passing no arguments. The documentation seems to disagree with this (`The lpCommandLine parameter can be NULL. In that case, the function uses the string pointed to by lpApplicationName as the command line.`) so I'd avoid doing this.

**Comment by AdmiralCurtiss** [comment]
tldr: The PR should be correct as-is. I'll also test `ShellExecuteW()` shortly, just so we know what's up there...

**Comment by AdmiralCurtiss** [comment]
```
#include <string>

#include <Windows.h>

int main(int argc, char** argv) {
	std::wstring unquoted = L"c:\\Test\\a dir\\a.exe";
	std::wstring quoted = L"\"" + unquoted + L"\"";
	ShellExecuteW(nullptr, nullptr, L"explorer.exe", unquoted.c_str(), nullptr, SW_SHOW);;
}
```

Launches the correct executable, regardless of if I pass it quoted or unquoted. So that should also be fine.

If I delete the executable it just opens Explorer (in My Documents), which is weird but at least not harmful.

**Comment by MayImilae** [comment]
It's been confirmed that this resolved the issue. https://forums.dolphin-emu.org/Thread-recent-update-triggering-this-error-on-auto-update?pid=529406#pid529406
